mono/metadata/object.c

```diff
@@ -7195,21 +7195,19 @@ mono_string_to_utf16 (MonoString *s)
 {
 	MONO_REQ_GC_UNSAFE_MODE;
 
-	char *as;
-
 	if (s == NULL)
 		return NULL;
 
-	as = (char *)g_malloc ((s->length * 2) + 2);
-	as [(s->length * 2)] = '\0';
-	as [(s->length * 2) + 1] = '\0';
+	int const length = s->length;
 
-	if (!s->length) {
-		return (gunichar2 *)(as);
-	}
+	gunichar2 * const as = (gunichar2 *)g_malloc ((length * 2) + 2);
+	as [length] = 0;
+
+	if (!length)
+		return as;
 	
-	memcpy (as, mono_string_chars(s), s->length * 2);
-	return (gunichar2 *)(as);
+	memcpy (as, mono_string_chars(s), length * 2);
```

Comment:
Same as https://github.com/mono/mono/pull/6267/files#diff-55205fbc5be02187cdfe6d77547ab70fR7203, you should replace that `2` by `sizeof(gunichar2)`.